packages/html-reporter/src/reportView.tsx

```diff
@@ -42,51 +43,75 @@ const testCaseRoutePredicate = (params: URLSearchParams) => params.has('testId')
-  const filteredStats = React.useMemo(() => computeStats(report?.json().files || [], filter), [report, filter]);
+
+  const filteredFiles = React.useMemo(() => {
+    const files = reportData?.files ?? [];
+
+    return files
+        .map(file => {
+          const tests = file.tests.filter(test => filter.matches(test));
+          return { ...file, tests };
+        })
+        .filter(file => file.tests.length !== 0);
+  }, [filter, reportData?.files]);
+
+  const filteredStats = React.useMemo(() => computeStats(filteredFiles), [filteredFiles]);
 
   return <div className='htmlreport vbox px-4 pb-4'>
     <main>
-      {report?.json() && <HeaderView stats={report.json().stats} filterText={filterText} setFilterText={setFilterText}></HeaderView>}
-      {report?.json().metadata && <MetadataView {...report?.json().metadata as Metainfo} />}
+      {reportData && <HeaderView stats={reportData.stats} filterText={filterText} setFilterText={setFilterText}></HeaderView>}
+      {reportData?.metadata && <MetadataView {...reportData.metadata as Metainfo} />}
       <Route predicate={testFilesRoutePredicate}>
         <TestFilesView
-          report={report?.json()}
+          report={reportData}
           filter={filter}
           expandedFiles={expandedFiles}
           setExpandedFiles={setExpandedFiles}
-          projectNames={report?.json().projectNames || []}
+          projectNames={reportData?.projectNames ?? []}
           filteredStats={filteredStats}
+          filteredFiles={filteredFiles}
         />
       </Route>
       <Route predicate={testCaseRoutePredicate}>
-        {!!report && <TestCaseViewLoader report={report}></TestCaseViewLoader>}
+        {!!report && (
+          <TestCaseViewLoader
+            report={report}
+            filteredFiles={filteredFiles}
+            filter={filter}
+          />
+        )}
       </Route>
     </main>
   </div>;
 };
 
 const TestCaseViewLoader: React.FC<{
   report: LoadedReport,
-}> = ({ report }) => {
-  const searchParams = new URLSearchParams(window.location.hash.slice(1));
+  filteredFiles: TestFileSummary[],
+  filter: Filter
+}> = ({ report, filteredFiles, filter }) => {
+  const searchParams = useSearchParams();
   const [test, setTest] = React.useState<TestCase | undefined>();
-  const testId = searchParams.get('testId');
+  const testId = searchParams.get('testId') ?? '';
   const anchor = (searchParams.get('anchor') || '') as 'video' | 'diff' | '';
   const run = +(searchParams.get('run') || '0');
 
   const testIdToFileIdMap = React.useMemo(() => {
     const map = new Map<string, string>();
-    for (const file of report.json().files) {
+    for (const file of filteredFiles) {
       for (const test of file.tests)
         map.set(test.testId, file.fileId);
     }
     return map;
-  }, [report]);
+  }, [filteredFiles]);
+
+  const { prevTestId, nextTestId } = getAdjacentTestIds(testIdToFileIdMap, testId);
```

Comment:
Computing adjacent tests is not natural to testcaseview, would be better to minimize it's awareness and only provide it with prev/next links that would navigate to the respective tests.

Reply:
With the current approach `TestCaseViewLoader` component creates a map with all test ids and is responsible for loading the current test. It seems to me like a reasonable place to compute tests adjacent to the current one, isn't it?

<img width="684" alt="Screenshot 2024-07-23 at 10 54 34" src="https://github.com/user-attachments/assets/cc931cda-5c0a-4707-9b0f-35ba37d94717">

\
I'm not sure that computing this in the parent component (`ReportView`) would be better. This component is also responsible for displaying the list of tests which does not need to know next and previous ids. 

<img width="760" alt="Screenshot 2024-07-23 at 11 38 59" src="https://github.com/user-attachments/assets/501e3163-3daf-4465-a321-4c6b2b1fa65c">

Perhaps I'm missing something and you have a completely different approach in mind. In this case please share it 🙂